ZBus: Multi-domain support

Trond-F-Christiansen opened the pull request:
This PR introduces multi-domain support for ZBus, enabling message forwarding between domains.

- **Shadow Channels**:
A reserved channel type used for observing channels originating in other domains.
- **Proxy Agents**:
Background processes that forward messages between domains. Handles message forwarding, receiving, acknowledgments, and publishing to shadow channels.
   - **IPC backend**: For multi domain between cores.
   - **UART backend**: For multi domain between devices.

Fixes #94857

rodrigopex reviewed (changes requested):
Thank you for your contribution. Here are some areas where we can improve:
1. We should try to implement the shadow channel without adding extra fields to the channel structure. Consider using the validator as a known function to check if the publisher is the proxy thread.
2. Avoid hardcoding proxy details into the code. Instead, retrieve device information from the DTS rather than specifying it directly.
3. Generally, regular channels should remain unchanged. Try to incorporate additional complexity within the proxy agent rather than in the user code. Perhaps making the proxy agents communicate with each other about the channels to be observed would be a good approach, and runtime observers can be helpful in this regard.

decsny commented on dts/bindings/zbus/zephyr,zbus-proxy-agent-base.yaml:
```diff
@@ -0,0 +1,48 @@
+# Copyright (c) 2025 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+description: |
+    Base properties for Zbus proxy agent proxy agents.
```
? double term?

decsny reviewed (dismissed):
I don't really like the idea of adding ZBUS as another pure piece of software configured by DT. Were there any alternatives considered.

rodrigopex commented on subsys/zbus/proxy_agent/Kconfig:
```diff
@@ -0,0 +1,54 @@
+config ZBUS_PROXY_AGENT_IPC
+	bool "ZBus proxy agent IPC backend"
+	depends on IPC_SERVICE
+	depends on CRC
```
Ditto.

rodrigopex commented on subsys/zbus/proxy_agent/Kconfig:
```diff
@@ -0,0 +1,54 @@
+	help
+	  Configures the priority of the ZBus proxy agent proxy agent thread.
+
+config ZBUS_PROXY_AGENT_UART
```
Are we able to select both at the same time? UART and IPC?

rodrigopex commented on subsys/zbus/proxy_agent/zbus_proxy_agent.c:
```diff
@@ -0,0 +1,788 @@
+		return config->tracking.ack_timeout_max_ms;
+	}
+
+	timeout = config->tracking.ack_timeout_start_ms * multiplier;
```
You may use math extras to remove that check: https://docs.zephyrproject.org/latest/doxygen/html/group__math__extras.html

rodrigopex commented on subsys/zbus/proxy_agent/zbus_proxy_agent.c:
```diff
@@ -0,0 +1,788 @@
+			config->backend.name, process_ret);
+		ret = schedule_ack(config, msg->id, ZBUS_PROXY_AGENT_MSG_TYPE_NACK);
+		if (ret < 0) {
+			LOG_ERR("Failed to schedule NACK for message %d: %d", msg->id, ret);
```
What happens in this scenario?

rodrigopex commented on subsys/zbus/proxy_agent/zbus_proxy_agent.c:
```diff
@@ -0,0 +1,788 @@
+
+static int zbus_proxy_agent_stop_tracking(struct zbus_proxy_agent_config *config, uint32_t msg_id)
+{
+	if (!config) {
```
ditto

rodrigopex commented on subsys/zbus/proxy_agent/zbus_proxy_agent.c:
```diff
@@ -0,0 +1,788 @@
+		return -EINVAL;
+	}
+
+	if (!config->tracking.tracking_msg_pool) {
```
ditto

rodrigopex commented on subsys/zbus/proxy_agent/zbus_proxy_agent.c:
```diff
@@ -0,0 +1,788 @@
+
+	buf = net_buf_alloc(config->tracking.tracking_msg_pool, K_NO_WAIT);
+
+	if (!buf) {
```
ditto

rodrigopex reviewed (changes requested):
There are some style change requests and a feature separation. Please take a look at that, and I can proceed with the review. Thank you for the massive contribution.

pdgendt commented:
> > throw it on the pile of neverending DT tech debt, w/e
> 
> @decsny, I could not understand why you are trying to do. Another [PR](https://github.com/zephyrproject-rtos/zephyr/pull/100085), extracted from this one, is currently under review.
> 
> This PR is significant, and it will not enter until it is mature enough. If you want to help, please do.

I think the main concern here is the (ab)use of Devicetree for describing software configuration. From [the docs](https://docs.zephyrproject.org/latest/build/dts/index.html):

> A devicetree is a hierarchical data structure primarily used to describe hardware.

There have been numerous discussions on whether this should be justified or not. A relevant proposal from @jukkar is to use yaml files for this, as can be seen in PR https://github.com/zephyrproject-rtos/zephyr/pull/97266

As a project, I do think we need to clarify the way forward.
_CC @carlescufi_

bjarki-andreasen reviewed (changes requested):
The feature is cool, we definitely could use it, but the devicetree is being misused, so its a nack until we get a better solution.

A backend which works on top of hardware, like a uart shell backend, or a driver using some bus as a transport layer, is not defined in the devicetree, nor should the zbus proxy agent, which is most definately not a piece of hardware.

Like a console or shell backend, specific hardware can be pointed to in the devicetree with a chosen node
```
chosen {
        zbus-proxy-agent-uart = &uart0;
};
```
but instance specific configurations specific to the software backend which uses the UART is defined in Kconfig, with the limitations that brings.

An application with many IPC backends would define the specific UARTs or IPC instances using zephyr,user
```
zephyr,user {
        foo-zbus-proxy-uart = <&uart0>;
        bar-zbus-proxy-ipc = <&ipc0>;
};
```
and the application would use these to statically define the backends/channels (this is very pseudo code)
```
ZBUS_PROXY_AGENT_UART_DEFINE(
        foo_proxy_agent,
        DT_PROP(DT_PATH(zephyr_user), foo_zbus_proxy_uart),
        1000, /* ack-timeout-start-ms */
        ... /* rest of configurations */
);

ZBUS_PROXY_AGENT_IPC_DEFINE(
        bar_proxy_agent,
        DT_PROP(DT_PATH(zephyr_user), bar_zbus_proxy_ipc),
        1000, /* ack-timeout-start-ms */
        ... /* rest of configurations */
);
```
These proxies will be very specific to applications, having the applications define them seems reasonable to me.

rodrigopex commented on subsys/zbus/proxy_agent/zbus_proxy_agent.c:
```diff
@@ -0,0 +1,793 @@
+
+static int handle_message_retry(struct zbus_proxy_agent_tracked_msg *data, uint32_t expected_msg_id)
+{
+	int ret;
```
ditto

nandojve commented:
Could the devicetree be expressed as below instead using the `zephyr,user` or `chosen`?

```c
/ {
	zbus {
		compatible = "zephyr,zbus";

		proxy-uart: proxy-uart {
			compatible = "zephyr,zbus-proxy-uart";
			channel = <&uart0>;
			ack-timeout = <1000>;
		};

		proxy-ipc: proxy-ipc {
			compatible = "zephyr,zbus-proxy-ipc";
			channel = <&ipc0>;
			ack-timeout = <1000>;
		};
	};
};
```

User can still define the node name and pick that in an static configuration. Inside ZBus each backend will look direct by the compatible. The ZBus group make the details more organized and clear.

bjarki-andreasen commented:
> > > Could the devicetree be expressed as below instead using the `zephyr,user` or `chosen`?
> > > User can still define the node name and pick that in an static configuration. Inside ZBus each backend will look direct by the compatible. The ZBus group make the details more organized and clear.
> > 
> > Again this would be moving software components to devicetree, which is what we're trying to avoid. As this is an application facing subsystem, we can express everything from C code, as suggested by @bjarki-andreasen
> 
> I do not share the same opinion. The presence of a name in the devicetree does not imply that the software is defined there. This area involves a significant overlap between hardware and software. From my perspective, a purely software component is one that can be defined entirely within Kconfig. As soon as the software depends on information from the devicetree, it enters this gray area.
> 
> The bindings in this proposal specify the communication backend channels, which are driver abstractions. These channels represent hardware communication paths between endpoints, and the bindings explicitly define which types of drivers should manage each instance. Therefore, this is not a purely software component, and I do not see any misuse of the devicetree.
> 
> If presenting these details in a structured way is considered “defining software in hardware,” then `chosen` and `zephyr,user` would effectively do the same. Consider examples such as USB class bindings, UART bridges, or modem socket offload. Some of these may be debatable, but they also reside within the same hardware–software gray area.
> 
> In my view, if we must work within this gray area because the software component cannot be fully defined in Kconfig, and the devicetree provides a clear mechanism for describing how these channels are defined, then not using it would be a missed opportunity. This is not the same as define how many instances of _philosophers_ the application should run.

chosen and zephyr,user do not define hardware nor software, they contain references to instances of hardware defined in the devicetree. They are the bare minimum required to portably bind an app to a board. There is no binding for them, they have no compat, they are just pointers.

nandojve commented:
@bjarki-andreasen , @pdgendt ,

I can apply the same reasoning you used for `chosen` and `zephyr,user` to this case. As you said, those nodes do not define hardware or software; they simply reference existing hardware instances already described in the devicetree. My proposal follows the same principle.

Defining a compatible group to associate hardware-related properties does not define software. It simply organizes hardware characteristics in a structured way. It even provides a clearer representation when multiple instances are present—which is a valid scenario—and still allows selecting a specific instance by its node name.

You are stating that my proposal defines software, but you have not identified where this software definition supposedly happens. If anything, the proposal describes hardware interfaces and their relationships, just like many other bindings that exist today.

So before we classify this as “defining software in the devicetree,” I think we need a clear explanation of which part of the binding is actually defining software rather than describing hardware-based communication structures.

bjarki-andreasen commented:
> I agree with @nandojve. Using `leds`, `pwmleds`, and `buttons` as examples, these are collections of data used in software. However, they only specify which GPIO pins correspond to an LED, PWM, or button, and what configurations should be applied in software. Nearly all the information stored in these groups is utilized during the software setup phase, not within the driver itself. What should be the difference in this case? A `zbus` group with the proxy agents mapping the devices and other information used in the setup process. In my view, we are discussing the same thing.
> 
> In this case, I consider @nandojve's solution appropriate. Maybe we can change the group name to `zbus-proxies`.
> 
> ```
> / {
> 	zbus-proxies {
> 		compatible = "zephyr,zbus-proxy";
> 
> 		proxy0: proxy-uart {
> 			compatible = "zephyr,zbus-proxy-uart";
> 			medium = <&uart0>;
> 			ack-timeout = <2000>;
> 		};
> 
> 		proxy1: proxy-ipc {
> 			compatible = "zephyr,zbus-proxy-ipc";
> 			medium = <&ipc0>;
> 			ack-timeout = <1000>;
> 		};
> 	};
> };
> ```

Difference is that leds and buttons are physical devices, on the physical board, with various device models (control interfaces), which point to the physical gpio and pwm instances which are connected to it.

zbus is not a device, nor is a zbus proxy. It has no physical control interface, it does not belong in the devicetree.

bjarki-andreasen reviewed (changes requested):
Looks really good, there is one lingering file otherwise its an ack from me

rodrigopex commented on subsys/zbus/proxy_agent/zbus_proxy_agent.c:
```diff
@@ -0,0 +1,764 @@
+		return -ENOSYS;
+	}
+
+	key = irq_lock();
```
Why IRQ lock here?

rodrigopex reviewed (changes requested):
I am worried about the separation of concerns here. Probably, some backend responsibilities are assigned to the Proxy agent. For example, in my view, retransmission should not be a concern for the "application" layer. That is a backend layer responsibility. Some backends (such as TCP) may not require retransmissions. 

The ability to manage different shadowed channels (in both directions) simultaneously made the complexity explode. You could keep that consistent and simple by enabling only one blocking transmission in one direction at a time. The Proxy Agent would receive messages from the backend when they are correct and confirmed (ACK). On the other hand, the Proxy Agent would send a message and wait for a backend response until the other domain ACKs it.

A good approach is to split the Proxy Agent into two parts. One receives updates, publishes to shadow channels, and ACKs the remote; the other monitors the original channels, sends updates to the other domain, and waits for ACKs. In this way, the backend would handle the duplex nature of the transmission medium (half- or full-duplex). 

I believe it would be more productive to have a meeting. Please contact me on Discord.

Trond-F-Christiansen committed:
tests: zbus: add multidomain proxy agent test suite

Add test suite for the zbus multidomain proxy agent
implementation.
adds a mock backend to enable isolated testing without requiring
actual hardware.

Signed-off-by: Trond F. Christiansen <trond.christiansen@nordicsemi.no>

Trond-F-Christiansen committed:
zbus: add IPC backend for zbus proxy agent

Implement IPC backend for ZBus proxy agent, enabling
message forwarding between domains over IPC endpoints.

Signed-off-by: Trond F. Christiansen <trond.christiansen@nordicsemi.no>

Trond-F-Christiansen committed:
samples: zbus: add ZBus proxy agent IPC sample

Add sample application demonstrating ZBus multidomain communication
over IPC between CPU cores. Shows practical usage of shadow channels
and proxy agents for inter-core message forwarding.

Signed-off-by: Trond F. Christiansen <trond.christiansen@nordicsemi.no>

Trond-F-Christiansen committed:
zbus: add multidomain proxy agent core

Add core implementation for zbus communication between domains using
pluggable backend transports.

Signed-off-by: Trond F. Christiansen <trond.christiansen@nordicsemi.no>

Trond-F-Christiansen committed:
doc: zbus: add documentation for multi-domain zbus

Add documentation for the multi-domain zbus feature

Signed-off-by: Trond F. Christiansen <trond.christiansen@nordicsemi.no>

rodrigopex commented on include/zephyr/zbus/proxy_agent/zbus_proxy_agent.h:
```diff
@@ -0,0 +1,356 @@
+	/** Resolved shadow channel that will receive the message */
+	const struct zbus_channel *chan;
+	/** Binary message data */
+	uint8_t message_data[CONFIG_ZBUS_PROXY_AGENT_MAX_MESSAGE_SIZE];
```
```suggestion
	uint8_t message[CONFIG_ZBUS_PROXY_AGENT_MAX_MESSAGE_SIZE];
```

rodrigopex reviewed (changes requested):
Great improvements in terms of simplicity, separation of concerns, and operation. I still have doubts about the agents using work queues. It seems better to use a dedicated thread since you are not sharing them. Another suggestion is in the name selection. Maybe be improved for clarity. There are several redundant `zbus_proxy_agent...`.

Trond-F-Christiansen committed:
tests: zbus: add multidomain IPC backend test suite

Add test suite for the zbus multidomain IPC backend.
Includes a mock IPC backend to enable isolated testing.

Signed-off-by: Trond F. Christiansen <trond.christiansen@nordicsemi.no>

rodrigopex commented on samples/subsys/zbus/proxy_agent/ipc/remote/src/main.c:
```diff
@@ -0,0 +1,97 @@
+int get_random_value(int min, int max)
+{
+	return min + (sys_rand32_get() % (max - min + 1));
+}
```
Remove this?

rodrigopex reviewed (changes requested):
Very good improvements. Missing minor details.

rodrigopex reviewed (dismissed):
The code seems ok. The feature is marked as experimental. Just make the checks pass.

pdgendt reviewed (commented):
Looks great! Couple of nits.

Trond-F-Christiansen commented on include/zephyr/zbus/proxy_agent/zbus_proxy_agent.h:
```diff
@@ -0,0 +1,347 @@
+#define ZBUS_PROXY_AGENT_DEFINE(_name, _type, _backend_dt_node)                                    \
+	_ZBUS_PROXY_AGENT_BACKEND_CONFIG_DEFINE(_name, _type, _backend_dt_node)                    \
+	K_THREAD_STACK_DEFINE(_name##_thread_stack,                                                \
+			      CONFIG_ZBUS_PROXY_AGENT_WORK_QUEUE_STACK_SIZE);                      \
```
Agreed, will have a look at it once the first version is in

aescolar commented on tests/subsys/zbus/proxy_agent/ipc_backend/src/main.c:
```diff
@@ -0,0 +1,324 @@
+	zassert_not_null(config, "Config should not be NULL");
+	zassert_equal(config->dev, DEVICE_DT_GET(FAKE_IPC_NODE),
+		      "Device should match the one from devicetree");
+	zassert_equal(config->ept_name, "test_agent_ipc_ept",
```
Issue in main here also, see fix in https://github.com/zephyrproject-rtos/zephyr/pull/106015
